src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java

```diff
@@ -3049,7 +3049,36 @@ ByteVector fromArray(VectorSpecies<Byte> species,
                                    byte[] a, int offset,
                                    int[] indexMap, int mapOffset) {
         ByteSpecies vsp = (ByteSpecies) species;
-        return vsp.vOp(n -> a[offset + indexMap[mapOffset + n]]);
+        IntVector.IntSpecies isp = IntVector.species(vsp.indexShape());
+        Objects.requireNonNull(a);
+        Objects.requireNonNull(indexMap);
+        Class<? extends ByteVector> vectorType = vsp.vectorType();
+
+        int loopIncr = 0;
+        VectorSpecies<Integer> lsp = null;
+
+        // Constant folding should sweep out following conditonal logic.
+        if (isp.length() > IntVector.SPECIES_PREFERRED.length()) {
+            lsp = IntVector.SPECIES_PREFERRED;
```

Comment:
I see, thanks for the explanation! It's different from SVE, which the index should be an int vector with the same length of the target vector.  We have to generate the index vectors. So the array length check to the given index map is necessary in java side.

Reply:
Intrinsic still receive the indexArray and corresponding species with matching lane count OR maximally supported integral species otherwise. Due to lane size differences b/w sub-word and integer it may never be possible to have compatible species.

Reply:
> Due to lane size differences b/w sub-word and integer it may never be possible to have compatible species.

That's true. So maybe we could just 1) make sure the index array is inside of the boundary (i.e. the length is at least the same with the vectors), and 2) generate the index vectors during intrinsification for some architectures that need it (i.e. ARM SVE). For now, since only x86 implement it, maybe the second one is not needed.